projects/storefrontlib/src/lib/cms-lib/product-carousel/product-carousel.component.ts

```diff
@@ -56,58 +36,36 @@ export class ProductCarouselComponent implements OnInit {
     this.setItems();
   }
 
-  /**
-   * Maps the item codes from CMS component to an array of `Product` observables.
-   */
   protected setItems(): void {
-    this.items$ = this.component.data$.pipe(
-      map(data => {
-        const productCodes = data.productCodes.split(' ');
-        return productCodes.map(code => this.productService.get(code));
-      })
-    );
+    this.items$ = this.productCarouselService.setItems();
   }
 
-  /**
-   * The number of items shown in the carousel can be calculated
-   * the standard implemenattions uses the element size to calculate
-   * the items that fit in the carousel.
-   * This method is called in `ngOnInit`.
-   */
   protected setItemSize(): void {
-    if (!this.window) {
-      this.itemSize$ = of(MAX_ITEM_SIZE);
-      return;
-    }
-    this.itemSize$ = fromEvent(this.window, 'resize').pipe(
-      map(() => (this.el.nativeElement as HTMLElement).clientWidth),
-      startWith((this.el.nativeElement as HTMLElement).clientWidth),
-      // avoid to much calls
-      debounceTime(100),
-      map((innerWidth: any) => {
-        const itemsPerPage = Math.round(innerWidth / MAX_WIDTH);
-        return itemsPerPage > 2 ? MAX_ITEM_SIZE : itemsPerPage;
-      }),
-      // only emit new size when the size changed
-      distinctUntilChanged()
+    this.itemSize$ = this.productCarouselService.setItemSize(
+      this.window,
+      this.el.nativeElement
     );
   }
 
-  prev(max): void {
-    this.setActiveItem(this.activeItem - max, max);
+  async prev(max) {
+    this.activeItem = await this.productCarouselService.setPreviousItemAsActive(
+      this.activeItem,
+      max
+    );
   }
 
-  next(max): void {
-    this.setActiveItem(this.activeItem + max, max);
+  async next(max) {
+    this.activeItem = await this.productCarouselService.setNextItemAsActive(
```

Comment:
if you do something like this:
```ts
await this.productCarouselService.delay(max);
this.activeItem = this.activeItem +1 // -1
```
these service methods will be no longer necessary

Reply:
refactored